Generalize initializing data loader

vshampor opened the pull request.

ljaljushkin commented on nncf/initialization.py:
```diff
@@ -105,14 +89,25 @@ def write(cls, msg):
+                    return isinstance(obj, torch.Tensor)
+
+                to_device_fn = partial(torch.Tensor.to, device=device)
+                args, kwargs = objwalk(args_and_kwargs_tuple, is_tensor, to_device_fn)
```
the same changes are needed in hessian_estimator.

vshampor replied:
Done

ljaljushkin commented on nncf/structures.py:
```diff
@@ -30,11 +30,13 @@ class QuantizationPrecisionInitArgs(NNCFExtraConfigStruct):
     """
 
-    def __init__(self, criterion: _Loss, data_loader: DataLoader):
+    def __init__(self, criterion: _Loss, data_loader: DataLoader, device: str = 'cuda'):
```
I think, better to use torch.device type

vshampor replied:
Probably `str` is what the users are more used to. Will also help generalize for TF.

vshampor commented:
@ljaljushkin note that I removed random transforms from the precision init test loaders. Somehow even this refactoring has upset the RNG call sequence and the previous reference traces no longer matched to the test traces. I updated the reference traces by first removing the random transforms on the current `develop` state, then updating this commit's reference traces with the test traces I got in the previous step.

vshampor commented:
Precommit failed due to CUDA OOM, because another test run was being executed concurrently on the same machine. Anyway, the latest update only affected the transformers integration patch and the associated readme, so I'll be merging this.
